platforms/posix/src/px4/common/console_buffer.cpp

```diff
@@ -0,0 +1,329 @@
+	void print(bool follow);
+
+	int size();
+
+	int read(char *buffer, int buffer_length, int *offset);
+
+	px4_sem_t _lock;
+	px4_sem_t _initialized;
+
+private:
+	void lock() { do {} while (px4_sem_wait(&_lock) != 0); }
+	void unlock() { px4_sem_post(&_lock); }
+
+	char _buffer[BOARD_CONSOLE_BUFFER_SIZE];
+	int _head{0};
+	int _tail{0};
+};
+
+void ConsoleBuffer::print(bool follow)
+{
+	const int buffer_length = 512;
+	char buffer[buffer_length];
+	int offset = -1;
+
+	do {
+		// print the full buffer or what's newly added
+		int total_size_read = 0;
+
+		do {
+			int read_size = read(buffer, buffer_length, &offset);
+
+			if (read_size <= 0) {
+				break;
+			}
+
+			if (::write(STDERR_FILENO, buffer, read_size) < 0) {
+				break;
+			}
+
+			if (read_size < buffer_length) {
+				break;
+			}
+
+			total_size_read += read_size;
+		} while (total_size_read < BOARD_CONSOLE_BUFFER_SIZE);
+
+
+		if (follow) {
+			usleep(10000);
+		}
+	} while (follow);
+}
+
+void ConsoleBuffer::write(const char *buffer, size_t len)
+{
+	lock();
+
+	for (size_t i = 0; i < len; ++i) {
+		_buffer[_tail] = buffer[i];
+		_tail = (_tail + 1) % BOARD_CONSOLE_BUFFER_SIZE;
+
+		if (_tail == _head) {
+			_head = (_head + 1) % BOARD_CONSOLE_BUFFER_SIZE;
+		}
+	}
+
+	unlock();
+}
+
+int ConsoleBuffer::size()
+{
+	lock();
+	int size;
+
+	if (_head <= _tail) {
+		size = _tail - _head;
+
+	} else {
+		size = BOARD_CONSOLE_BUFFER_SIZE - (_head - _tail);
+	}
+
+	unlock();
+	return size;
+}
+
+int ConsoleBuffer::read(char *buffer, int buffer_length, int *offset)
+{
+	lock();
+
+	if (*offset == -1) {
+		*offset = _head;
+	}
+
+	int size = 0;
+
+	if (*offset < _tail) {
+		size = _tail - *offset;
+
+		if (size > buffer_length) {
+			size = buffer_length;
+		}
+
+		memcpy(buffer, _buffer + *offset, size);
+
+	} else if (_tail < *offset) {
+		size = BOARD_CONSOLE_BUFFER_SIZE - *offset;
+
+		if (size > buffer_length) {
+			size = buffer_length;
+		}
+
+		memcpy(buffer, _buffer + *offset, size);
+		buffer += size;
+		buffer_length -= size;
+
+		int size_secondary = _tail;
+
+		if (size_secondary > buffer_length) {
+			size_secondary = buffer_length;
+		}
+
+		if (size_secondary > 0) {
+			memcpy(buffer, _buffer, size_secondary);
+			size += size_secondary;
+		}
+	}
+
+	unlock();
+	*offset = (*offset + size) % BOARD_CONSOLE_BUFFER_SIZE;
+	return size;
+}
+
+static ConsoleBuffer g_console_buffer;
+
+int px4_console_buffer_thread(int argc, char *argv[])
+{
+	int ret = PX4_OK;
+	int orig_stdout_fd = dup(STDOUT_FILENO);
+
+	// Check if the console_buf named pipe already exists. If so delete it.
+	struct stat stats;
+
+	if (stat(CONSOLE_BUFFER_DEVICE, &stats) < 0) {
+
+		if (errno != ENOENT) {
+			perror("Console buffer stat failed");
+			ret = PX4_ERROR;
+			goto out;
+		}
+
+	} else  if (unlink(CONSOLE_BUFFER_DEVICE) < 0) {
+
+		perror("Console buffer unlink failed");
+		ret = PX4_ERROR;
+		goto out;
+	}
+
+	// Create console buffer named pipe
+	if ((mkfifo(CONSOLE_BUFFER_DEVICE, 0666)) < 0) {
```

Comment:
Can you make this an anonymous pipe? Otherwise it will create problems with multi-instances.

Reply:
The disadvantage of using px4_log_modulename to capture the output is that a significant amount of the PX4 code base bypasses the logging mechanism relying directly on printf() to provide user feedback.  This feedback is important to capture.  uavcan is an example.  In the long term moving to the unified logging mechanism, away from printf() style feedback, makes sense to me.  This would allow console_buffer functionality to be easily added to the existing logging functionality.  However, this requires touching a significant amount of code that has been tested over many years.  Capturing fd 1 is a short term solution.  Am I misinterpreting your px4_log_modulename suggestion?  Any thoughts or comments?

Reply:
> significant amount of the PX4 code base bypasses the logging mechanism relying directly on printf() to provide user feedback

Where do you see that? We introduced `PX4_INFO_RAW` for that, and it might not be used everywhere yet. But all printf's that you see should be moved over to that. uavcan is one of those cases.

Reply:
The current code base has printf() sprinkled in many areas (uavcan, mavlink, ekf2, microRTPS, some drivers, main.cpp, pxh.cpp, ...).  I didn't mean to imply that the logging mechanism isn't used extensively, or the main method, but rather printf()s are sprinkled in to provide status or error messages in a number of areas.  I agree the logging mechanism is already in place and should be used.  I'll close this pull request, and I'll open a separate one with some proposed changes the existing logging functionality.  Hopefully, over a reasonably short timeframe, the code base will move away from printf()s making it unnecessary to capture fd 1.